File: libraries/KMPKtorCoinbase/shared/src/commonMain/kotlin/dev/subfly/kmpktorcoinbase/featureHome/domain/usecase/GetAllCoinsUseCase.kt
package dev.subfly.kmpktorcoinbase.featureHome.domain.usecase

import dev.subfly.kmpktorcoinbase.core.data.mapper.toCoinHomeModel
import dev.subfly.kmpktorcoinbase.core.exception.InvalidIdException
import dev.subfly.kmpktorcoinbase.core.model.CoinHomeModel
import dev.subfly.kmpktorcoinbase.core.util.Resource
import dev.subfly.kmpktorcoinbase.featureHome.domain.dao.CoinPaprikaHomeDao
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class GetAllCoinsUseCase(private val repository: CoinPaprikaHomeDao) {
    suspend fun fetchData(): Flow<Resource<List<CoinHomeModel>>> = flow {
        emit(Resource.Loading)
        try {
            val coinsFetched = repository
                .getAllCoins()
                .map {
                    it.toCoinHomeModel()
                }
            emit(Resource.Success(coinsFetched))
        } catch (e: InvalidIdException) {
            emit(Resource.Error(e.message))
        } catch (e: Exception) {
            // General exception handling for Ktor, no deep dive, sorry
            e.message?.let { errorMessage ->
                emit(Resource.Error(errorMessage))
            }
        }
    }
}

File: ios/BaseiOSUIKit/BaseiOSUIKit/FeatureHome/Domain/UseCase/GetAllCoinsUseCase.swift
//
//  GetAllCoinsUseCase.swift
//  BaseiOSUIKit
//
//  Created by Ali Taha on 5.12.2023.
//

import Foundation
import Combine

class GetAllCoinsUseCase {
    private let repository: CoinPaprikaHomeDao
    
    var resource: CurrentValueSubject<Resource<[CoinHomeModel]>, Never>

    init(repository: CoinPaprikaHomeDao) {
        self.repository = repository
        self.resource = CurrentValueSubject(.loading)
    }

    func fetchData() async {
        do {
            resource.send(.loading)
            let coinsFetched = try await repository.getAllCoins().map { try $0.toCoinHomeModel() }
            resource.send(.success(data: coinsFetched))
        } catch let error as InvalidUrlException {
            resource.send(.error(message: error.message))
        } catch let error as InvalidIdException {
            resource.send(.error(message: error.message))
        } catch let error as InvalidResponseException {
            resource.send(.error(message: error.message))
        } catch {
            resource.send(.error(message: error.localizedDescription))
        }
    }
}
